cabal-install/src/Distribution/Client/CmdGenBounds.hs

```diff
@@ -0,0 +1,256 @@
+
+    -- Process each package to find the ones needing bounds
+    let boundsActions = concatMap genBoundsActionForPkg localPkgs
+
+    if (any isBoundsNeeded boundsActions)
+      then do
+        notice verbosity boundsNeededMsg
+        mapM_ (renderBoundsResult verbosity) boundsActions
+      else notice verbosity "All bounds up-to-date"
+
+data GenBoundsResult = GenBoundsResult PackageIdentifier ComponentTarget (Maybe [PackageIdentifier])
+
+isBoundsNeeded :: GenBoundsResult -> Bool
+isBoundsNeeded (GenBoundsResult _ _ Nothing) = False
+isBoundsNeeded _ = True
+
+renderBoundsResult :: Verbosity -> GenBoundsResult -> IO ()
+renderBoundsResult verbosity (GenBoundsResult pid tgt bounds) =
+  case bounds of
+    Nothing ->
+      notice
+        verbosity
+        ("Congratulations, all dependencies for " ++ prettyShow (packageName pid) ++ ":" ++ showComponentTarget pid tgt ++ " have upper bounds!")
+    Just pkgBounds -> do
+      notice verbosity $
+        "For component " ++ prettyShow (pkgName pid) ++ ":" ++ showComponentTarget pid tgt ++ ":"
+      let padTo = maximum $ map (length . unPackageName . packageName) pkgBounds
+      traverse_ (notice verbosity . (++ ",") . showBounds padTo) pkgBounds
+
+-- | Process a single BuildInfo to identify and report missing upper bounds
+getBoundsForComponent
+  :: ComponentTarget
+  -> ElaboratedConfiguredPackage
+  -> Map.Map ComponentId PackageIdentifier
+  -> GenBoundsResult
+getBoundsForComponent tgt pkg pkgVersionMap =
+  if null needBounds
+    then boundsResult Nothing
+    else -- All the things we depend on.
+
+      let componentDeps = elabLibDependencies pkg
+          -- Match these up to package names, this is a list of Package name to versions.
+          -- Now just match that up with what the user wrote in the build-depends section.
+          depsWithVersions = mapMaybe (\cid -> Map.lookup (confInstId $ fst cid) pkgVersionMap) componentDeps
+          isNeeded = hasElem needBounds . packageName
+       in boundsResult (Just (filter isNeeded depsWithVersions))
+  where
+    pd = elabPkgDescription pkg
+    -- Extract the build-depends for the right part of the cabal file.
+    bi = buildInfoForTarget pd tgt
+
+    -- We need to generate bounds if
+    -- \* the dependency does not have an upper bound
+    -- \* the dependency is not the same package as the one we are processing
+    boundFilter dep =
+      (not (hasUpperBound (depVerRange dep)))
+        && packageName pd /= depPkgName dep
+
+    -- The dependencies that need bounds.
+    needBounds = map depPkgName $ filter boundFilter $ targetBuildDepends bi
+
+    boundsResult = GenBoundsResult (packageId pkg) tgt
+
+buildInfoForTarget :: PackageDescription -> ComponentTarget -> BuildInfo
+buildInfoForTarget pd (ComponentTarget cname _) = componentBuildInfo $ getComponent pd cname
+
+-- | This defines what a 'TargetSelector' means for the @gen-bounds@ command.
+-- Copy of selectPackageTargets from CmdBuild.hs
```

Comment:
This should probably be extracted and put in a common place, rather than copied.